napari-workshops/notebooks/image_warping.md

```diff
@@ -0,0 +1,211 @@
+
+Mouse callbacks are essentially a *generator*, making use of `yield`s to split up the execution into separate code for handling the mouse press, the mouse drag, and the mouse release events. The dummy function below describes the structure of a mouse callback.
+
+```{code-cell} ipython3
+def some_mouse_callback(interacted_layer, event):
+    # here, we put any code that needs to be executed when the mouse is first pressed
+    print("Mouse has been pressed")
+    # once this code is executed, we yield to the calling function to execute any other 
+    # code that needs to run when the mouse is pressed
+    yield
+
+    # now, we handle the mouse drag. Because the drag can last indefinitely, and we
+    # don't want our viewer to freeze while we drag, we put our dragging code into
+    # a while loop and yield after each iteration
+    while event.type == 'mouse_move':
+        print(f"Mouse is being dragged. Current mouse position is {event.position}.")
+        yield
+
+    # the while loop will end when the mouse is no longer being dragged i.e. when
+    # the mouse button is released
+    print("Mouse has been released!")
+```
+
+The callback above doesn't do anything useful, but let's see what happens if we attach it to the image layer. Run the code below, select the `checkerboard` image layer in the viewer, and click and drag the image around. You should see the information printed in the terminal. Note that if some other layer is selected, you don't get anything printed. Also note that if you just make a single click, we don't get the information about the drag. Looks like our structure is correct!
+
+```{code-cell} ipython3
+checkerboard_image_layer.mouse_drag_callbacks.append(some_mouse_callback)
+```
+
+Remember to disconnect your callback! For mouse callbacks, we call `remove` on the `mouse_drag_callbacks` and pass in our function reference.
+
+```{code-cell} ipython3
+checkerboard_image_layer.mouse_drag_callbacks.remove(some_mouse_callback)
+```
+
+Let's edit our callback to pass the correct information to the `warp` function when the mouse is dragged.
+
+To pass the right information to the callback, we inspect the `selected_data` attribute of our `points_layer`, which contains the indices of all currently selected points. This example only works when one point is selected, so we just grab the first one. We make a copy of the `dst` points before updating the coordinates of the selected point to the current mouse position, to avoid messing with the layer data itself. Finally, we call the same warping function as in our layer event example.
```

Comment:
```suggestion
To pass the right information to the callback, we inspect the `selected_data` attribute of our `points_layer`, which contains the indices of all currently selected points. This example only works when one point is selected, so we just grab the last one. We make a copy of the `dst` points before updating the coordinates of the selected point to the current mouse position, to avoid messing with the layer data itself. Finally, we call the same warping function as in our layer event example.
```
You use last in the code :)